### PR TITLE
feat (core \log\db): areatrigger.dbc, sAreaTriggerStore, areatrigger_dbc

### PR DESCRIPTION
**Pull request by acidmanifesto**
Alrighty so I am not certain who thought it would be wonderful to not use the areatrigger.dbc file but in its place use a areatrigger table with 1217 entries where the dbc has 1220 entries. I am not doing a comparison to figure out the missing shit. It is excessive and not necessary. 

## Changes Proposed:
-  get rid of areatrigger table 
-  use and implement areatrigger.dbc
- implement areatrigger_dbc table
- Implement sAreaTriggerStore

## Issues Addressed:

- Closes areatrigger.dbc not being used
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/15432

## Tests Performed:

- builds
- runs

## How to Test the Changes:

1. This is a heavily global core wide change.
2. Test affect boss scripts
3. test any areas of areatriggers.
4. Run a affect BG
5. SMART_SCRIPT_TYPE_AREATRIGGER needs to be tested
6. battlegroundEY and WS need to be tested.
7. HandleGoTriggerCommand gm cmd needs to be tested
8. various bosses needs to be tested that were adjusted in the file changes.
9. Continue betting pool with how long it will be before review and merge.  Currently it is between 6 months and 10 months.

Restore for original world areatrigger table that gets dropped with this PR:
https://gist.github.com/acidmanifesto/b5bffecf2446c707d386af23057d8cd9

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by acidmanifesto** [comment]
Pr is ready for testing. Just slapping sensitive wip so we can address any issues any testers will find.

**Comment by acidmanifesto** [comment]
current failure is with lua module, pr for that to be done another time

image

**Comment by 55Honey** [comment]
Test with https://github.com/azerothcore/mod-eluna/pull/93

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -4313,20 +4313,25 @@ void SmartScript::OnInitialize(WorldObject* obj, AreaTrigger const* at)
...
+                    trigger = at;
+                    atPlayer = obj->ToPlayer();
+                    LOG_DEBUG("sql.sql", "SmartScript::OnInitialize: source is AreaTrigger {}, triggered by player {}", trigger->ID, atPlayer->GetGUID().ToString().c_str());
+                }
+                else
+                    LOG_ERROR("sql.sql", "SmartScript::OnInitialize: !WARNING! Player TypeID is only allowed for AreaTriggers");
+                break;
             default:
-                LOG_ERROR("scripts.ai.sai", "SmartScript::OnInitialize: Unhandled TypeID !WARNING!");
+                LOG_ERROR("sql.sql", "SmartScript::OnInitialize: Unhandled TypeID !WARNING!");
```
Comment (Kitzunu) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:4328]: why do we not keep this as scripts.ai.sai?

**Review by Kitzunu — APPROVED** [review]
All in all looks good.

Do we not want to be able to add Areatriggers if they were not to exist? As you are removing the table fully?

File: src/server/game/AI/SmartScripts/SmartScript.cpp (hunk 1 above)
Comment (acidmanifesto) [reply src/server/game/AI/SmartScripts/SmartScript.cpp:4328]: > why do we not keep this as scripts.ai.sai?

We have no conf defaul for the logging while all the others are sql.sql
Best we keep it consistent.

**Comment by acidmanifesto** [comment]
> All in all looks good.
> 
> Do we not want to be able to add Areatriggers if they were not to exist? As you are removing the table fully?

The asshat who added the table only literally did 1217 lines/ 99% of the dbc. We do not have it all extracted from the dbc to the table which is 1220 lines. We do not even use the dbc for some reason. This pr. We yeet out the shit hard coded table, and use the propper dbc with a areatrigger_dbc over ride table to apply custom buttstuff.

**Comment by Nyeriah** [comment]
Let's not use rude language when referring to our contributors please

**Comment by Nyeriah** [comment]
FYI https://github.com/azerothcore/azerothcore-wotlk/commit/05cd8c8c61947fc46a4ea7ceebf227624da698e4

**Comment by acidmanifesto** [comment]
> FYI https://github.com/azerothcore/azerothcore-wotlk/commit/05cd8c8c61947fc46a4ea7ceebf227624da698e4

Humor me. 
Why move it to db? 
Why not all the the entries from the dbc being exported? Why exactly would we benefit with it being soley in the db adding more to world being hit on? 
Just... why? Was it to see if it could be done?
There is zero benefits that i see in having this still soley stored in the world db instead of utilizing the dbc file in full.

**Comment by Nyeriah** [comment]
> > FYI [05cd8c8](https://github.com/azerothcore/azerothcore-wotlk/commit/05cd8c8c61947fc46a4ea7ceebf227624da698e4)
> 
> Humor me. Why move it to db? Why not all the the entries from the dbc being exported? Why exactly would we benefit with it being soley in the db adding more to world being hit on? Just... why? Was it to see if it could be done? There is zero benefits that i see in having this still soley stored in the world db instead of utilizing the dbc file in full.

That's a question for @Yehonal and @FrancescoBorzi as I was not part of this project when that happened and apparently there was a bounty awarded for this so they should probably be consulted before you revert it

Anyways, as @Kitzunu mentioned I believe we should still allow the database override/editing of data through database, like spell_dbc

**Comment by acidmanifesto** [comment]
> 
> Anyways, as @Kitzunu mentioned I believe we should still allow the database override/editing of data through database, like spell_dbc

Thats what this pr allows. At current we do not touch the more full areatrigger.dbc file at all. It all soley reads from the incomplete areatrigger table.

**Comment by acidmanifesto** [comment]
> All in all looks good.
> 
> Do we not want to be able to add Areatriggers if they were not to exist? As you are removing the table fully?

Incase someone is wondering this. The current areatrigger table does gets dropped with this PR. Its a partial extraction of the dbc and incomplete with azerothcore currently. We drop it and use the dbc directly while offering the areatrigger_dbc table as a override which will accept custom AT.

**Comment by Yehonal** [comment]
> > FYI [05cd8c8](https://github.com/azerothcore/azerothcore-wotlk/commit/05cd8c8c61947fc46a4ea7ceebf227624da698e4)
> 
> Humor me. Why move it to db? Why not all the the entries from the dbc being exported? Why exactly would we benefit with it being soley in the db adding more to world being hit on? Just... why? Was it to see if it could be done? There is zero benefits that i see in having this still soley stored in the world db instead of utilizing the dbc file in full.

I'm not sure about the areatrigger case, honestly. The system of reading dbc from the database is only needed to allow you to implement your data from scratch or override the dbc ones without needing to edit dbc files (which are proprietary)

More info here: 

* https://github.com/azerothcore/azerothcore-wotlk/issues/584#issue-253085904
* https://github.com/azerothcore/azerothcore-wotlk/discussions/5307

However, this PR actually doesn't break that system so, unless it breaks other thing, it's good to go for me

**Comment by Roxtedy** [comment]
> > > FYI [05cd8c8](https://github.com/azerothcore/azerothcore-wotlk/commit/05cd8c8c61947fc46a4ea7ceebf227624da698e4)
> > 
> > 
> > Humor me. Why move it to db? Why not all the the entries from the dbc being exported? Why exactly would we benefit with it being soley in the db adding more to world being hit on? Just... why? Was it to see if it could be done? There is zero benefits that i see in having this still soley stored in the world db instead of utilizing the dbc file in full.
> 
> I'm not sure about the areatrigger case, honestly. The system of reading dbc from the database is only needed to allow you to implement your data from scratch or override the dbc ones without needing to edit dbc files (which are proprietary)
> 
> More info here:
> 
> * [[IDEA] Converting dbc to sql and reading from there [$170 awarded] #584 (comment)](https://github.com/azerothcore/azerothcore-wotlk/issues/584#issue-253085904)
> * [[SPELLS] why they must be stored and maintained in database (almost all of them) #5307](https://github.com/azerothcore/azerothcore-wotlk/discussions/5307)
> 
> However, this PR actually doesn't break that system so, unless it breaks other thing, it's good to go for me

I think so too. If this PR doesnt break anything, it should be

**Comment by acidmanifesto** [comment]
https://github.com/azerothcore/azerothcore-wotlk/issues/15432

**Comment by Kitzunu** [comment]
Can merge, or?

**Comment by acidmanifesto** [comment]
> Can merge, or?

Would need independent testing and also need to ensure that pr is on standby with the custom module its failing on.

**Comment by Kitzunu** [comment]
@azerothcore/testers Can someone take care of it, thanks =)

**Comment by acidmanifesto** [comment]
Just fyi: pr from @55Honey to fix custom build failure
https://github.com/azerothcore/mod-eluna/pull/93

**Comment by acidmanifesto** [comment]
At the moment the only failure in the workflow is with Lua Module in the custom module workflow.
image
Where we have a redefinition typedef AreaTrigger AreaTriggerEntry; . a solution pr is already put in by Honey to correct it after merge . Other then that, this works fully as intended. At this point we only need independent testing for this PR outside of my reports of it working.
The issue this PR addresses is we are not using the DBC for area triggers. We some how got it soley being ready from a SQL table and not all the entries are present there from the DBC which would lead to stuff obviously being broken that may or not be reported.  The purpose of this pr was to utlize the dbc which it isnt currently. a quick lesson, which is copy pasta from my post in discord. DBC is a DataBase Container from what i remember in some chitchat in some of the reverse engineering rooms back then. Could be database client, as it does come from a client, didnt really care honestly, we just say dbc and leave it at that, It can improve performance. By storing data in a binary format, the DBC store can improve the performance of the game's servers. This is because the binary format is more efficient than the text format used by the SQL database. The DBC store is used to store game data that is not frequently changed, such as item definitions, creature definitions, and zone definitions. The data is consider blizzlike in the fullest. It makes it easy to update the game's data, as changes can be made to the DBC files without having to modify the game's code in most cases although certain things will require mpq modifications. Currently We have it hard coded while also... listing wrong column names for the dbc in the enums... this corrects it, and no longer leaves it hard coded on a source level. Load times are imrpoved by a few ms for server started up and for character creations as well from viewed in debugger. Pr is already tested in full from a former tester and review is marked as approved.
However on the 29th i will be no longer maintaining the pr and will be closed then. Anyone is free to take it over if that comes by.

**Comment by Kitzunu** [comment]
If you have tested it, then I am good to merge it

**Comment by acidmanifesto** [comment]
> If you have tested it, then I am good to merge it

It was tested on my end however i am not going to over a 1000 spots to test each one and would like independent testing to be done.

**Pull request closed without merging**